Add AKS recover cluster objects from backup (#8)
appuio/rancher-docs: docs/modules/ROOT/pages/how-tos/aks_recover_from_backup.adoc

> = AKS Recover cluster objects from Azure Blob Storage backup

Reviewer: AKS can probably be omitted from the page title

```suggestion
= Recover cluster objects from Azure Blob Storage backup
```

== Prerequisites

* Executables used in this guide:
** `kubectl`

> ** `yq`

Reviewer: Specify the required yq version. yq 3.x and 4.x have completely different invocations.

Author: What a night mare! I've got [kislyuk/yq](https://github.com/kislyuk/yq) via pip.
```
$ yq --version
yq 2.12.2
```
Is there an "official" version we are going to use? I mean pip is probably you get everywhere to work.

Reviewer: Required yq version still needs to be addressed.

Reviewer 2: The one we use at the cloudscale and exoscale cluster installation guide is this one in Go: https://github.com/mikefarah/yq

> ** `yq`

Reviewer 2: I'm guessing you are using v4 as well?
Some distros ship v3 or older, v4 is completely different though. I think we should really clarify this as all the commands could fail with a mismatching yq version.
```suggestion
** `yq` https://mikefarah.gitbook.io/yq[yq YAML processor] (version 4 or higher)
```

== General procedure

1. Collect configuration for restic
2. Identify and retrieve restic snapshot
3. Extract files containing the desired objects and prepare them
4. Apply objects to the cluster

== Collect restic configuration

Restic requires the environment variables `RESTIC_REPOSITORY`, `RESTIC_PASSWORD`, `AZURE_ACCOUNT_NAME` and `AZURE_ACCOUNT_KEY` to be set.
They can be obtained from the target cluster itself.
They can also be obtained from the cluster catalog and Vault.

=== Obtaining restic configuration from cluster

[source,console]
----
export RESTIC_REPOSITORY=$(kubectl -n syn-cluster-backup get schedule objects -o jsonpath='azure:{.spec.backend.s3.bucket}:/')
export RESTIC_PASSWORD=$(kubectl -n syn-cluster-backup get secret objects-backup-password -o jsonpath='{.data.password}' | base64 --decode)
export AZURE_ACCOUNT_NAME=$(kubectl get secret admin-credentials-minio-k8up -n syn-minio -o jsonpath='{.data.accesskey}' | base64 --decode)
export AZURE_ACCOUNT_KEY=$(kubectl get secret admin-credentials-minio-k8up -n syn-minio -o jsonpath='{.data.secretkey}' | base64 --decode)
----

=== Obtaining restic configuration from catalog and vault

> 1. Obtain the repository URL of the clusters catalog and export it to `REPO_URL`

Reviewer: ```suggestion
1. Obtain the repository URL of the cluster's catalog and export it to `REPO_URL`
```

or

```suggestion
1. Obtain the repository URL of the cluster catalog and export it to `REPO_URL`
```

+
Get the URL from https://control.vshn.net/syn/lieutenantclusters.
Alternatively, get it from the Lieutenant API or the Kubernetes API Lieutenant is running on.
+
.Use the Lieutenant API
[source,console]
----
REPO_URL=$(curl -sH "${LIEUTENANT_AUTH}" "https://${LIEUTENANT_URL}/clusters/${CLUSTER_ID}" | jq -r .gitRepo.url)
----
+
.Use the Kubernetes API
[source,console]
----
REPO_URL=$(kubectl -n ${LIEUTENANT_NS} get cluster -o jsonpath='{.spec.gitRepoURL}' ${CLUSTER_ID})
----

2. Download and extract the cluster catalog
+
[source,console]
----
mkdir catalog

> git archive --remote ${REPO_URL} master | tar -xC catalog

Reviewer: What's the reasoning for preferring `git archive` over a regular `git clone` here? I don't really mind either way, just curious.

Author: Copy pasta from the OCP4 variant of this documentation appuio/openshift4-docs@3212ef5f44c04be8e65e4d3c15bec4b907c8ada0 Didn't reflect on that.

----

> 3. Login to vault

Reviewer: ```suggestion
3. Login to Vault
```

+
[source,console]
----
export VAULT_ADDR=https://vault-prod.syn.vshn.net
vault login -method=ldap username=<your username>
----

4. Export restic configuration
+
[source,console]
----

> export RESTIC_REPOSITORY=azure:$(yq -r '. | select(.kind == "Schedule").spec.backend.s3.bucket' catalog/manifests/cluster-backup/05_schedule.yaml):/

Reviewer: I assume this works with yq 3.x. For yq 4.x (4.9.6), I get:

```
$ yq -r '. | select(.kind == "Schedule").spec.backend.s3.bucket' catalog/manifests/cluster-backup/05_schedule.yaml 
Error: unknown command "catalog/manifests/cluster-backup/05_schedule.yaml" for "yq"
Run 'yq --help' for usage.
```

The correct invocation for yq 4.x is

```
yq e '. | select(.kind == "Schedule").spec.backend.s3.bucket' catalog/manifests/cluster-backup/05_schedule.yaml
```

PASSWORD_KEY="$(yq -r '. | select(.kind == "Secret" and .metadata.name == "objects-backup-password").stringData.password' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"
export RESTIC_PASSWORD=$(vault kv get -format json "clusters/kv/${PASSWORD_KEY%/*}" | jq -r ".data.data.${PASSWORD_KEY##*/}")

ID_KEY="$(yq -r '. | select(.kind == "Secret" and .metadata.name == "object-backup-s3-credentials").stringData.username' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"
export AZURE_ACCOUNT_NAME=$(vault kv get -format json "clusters/kv/${ID_KEY%/*}" | jq -r ".data.data.${ID_KEY##*/}")

SECRET_KEY="$(yq -r '. | select(.kind == "Secret" and .metadata.name == "object-backup-s3-credentials").stringData.password' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"
export AZURE_ACCOUNT_KEY=$(vault kv get -format json "clusters/kv/${SECRET_KEY%/*}" | jq -r ".data.data.${SECRET_KEY##*/}")
----

== Identify and retrieve snapshot

> 1. List the available snapshots.
>    Identify the one you do want to restore.
>    Take note of its `ID`.
> +
> [source,console]
> ----
> restic snapshots
> ----
> 
> 2. Retrieve the backup archive
> +

Reviewer 2: You can achieve automatic numbering by simply having the dot without the number.

Also, the "console" style doesn't exist in asciidoctor (that's more Markdown style), might as well leave it empty. Use bash instead, as that will render a nicer output IMO.

there are multiple occurrences of my suggestion, but I'm not going to mark each now.
```suggestion
. List the available snapshots.
   Identify the one you do want to restore.
   Take note of its `ID`.
+
[source,bash]
----
restic snapshots
----

. Retrieve the backup archive
+
```

Reviewer: As far as I understand the `console` style does exist for our Antora theme, and allows you to copy-paste commands when they're given with a shell prompt. I've observed that leading `#` or `$` are not selectable for blocks styled with `console`.

However, if you're giving commands without shell prompts, the style doesn't really matter.

[source,bash]
----
restic restore <ID> --target cluster-backup-object-restore-$(date +%F)
----

== Extract and prepare files

. List files in the backup.
   Take note of the path containing the required files.
+
[source,bash]
----

> tar tvf syn-cluster-backup-object-dumper.tar.gz

Reviewer: After checking the documentation for `restic`, this command needs to be executed in the directory created by restic in the previous step, since `--target` specifies the directory to which to extract the snapshot.

Author: Added the `cd`ahead to switch into the target directory.

> tar tvf syn-cluster-backup-object-dumper.tar.gz

Reviewer: Does the `--target` argument to `restic restore` mean that we need to change directory for this command to work?

= Recover cluster objects from Azure Blob Storage backup

== Prerequisites

* Executables used in this guide:
** `kubectl`

> ** https://github.com/kislyuk/yq[yq: Command-line YAML/XML/TOML processor] installable via `pip3 install yq`

Reviewer: This is a different `yq` than the one we generally use (https://github.com/mikefarah/yq), and seems to have nothing in common with the one by mikefarah which our documentation commonly refers to.

I think we should keep our documentation referring to only one `yq` implementation. Since there's already multiple documentations that refer to https://github.com/mikefarah/yq, I'd suggest going with that one, and requiring v4 or newer.

I've added comments below to update the commands to work with that yq.

Author: Applied.

----

=== Obtaining restic configuration from catalog and vault

. Obtain the repository URL of the cluster's catalog and export it to `REPO_URL`
+
Get the URL from https://control.vshn.net/syn/lieutenantclusters.
Alternatively, get it from the Lieutenant API or the Kubernetes API Lieutenant is running on.
+
.Use the Lieutenant API
[source,bash]
----
REPO_URL=$(curl -sH "${LIEUTENANT_AUTH}" "https://${LIEUTENANT_URL}/clusters/${CLUSTER_ID}" | jq -r .gitRepo.url)
----
+
.Use the Kubernetes API
[source,bash]
----
REPO_URL=$(kubectl -n ${LIEUTENANT_NS} get cluster -o jsonpath='{.spec.gitRepoURL}' ${CLUSTER_ID})
----

. Download and extract the cluster catalog
+
[source,bash]
----
mkdir catalog
git archive --remote ${REPO_URL} master | tar -xC catalog
----

. Login to Vault
+
[source,bash]
----
export VAULT_ADDR=https://vault-prod.syn.vshn.net
vault login -method=ldap username=<your username>
----

. Export restic configuration
+
[source,bash]
----
export RESTIC_REPOSITORY=azure:$(yq -r '. | select(.kind == "Schedule").spec.backend.s3.bucket' catalog/manifests/cluster-backup/05_schedule.yaml):/

> PASSWORD_KEY="$(yq -r '. | select(.kind == "Secret" and .metadata.name == "objects-backup-password").stringData.password' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"

Reviewer: ```suggestion
PASSWORD_KEY="$(yq e '. | select(.kind == "Secret" and .metadata.name == "objects-backup-password").stringData.password' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"
```

export RESTIC_PASSWORD=$(vault kv get -format json "clusters/kv/${PASSWORD_KEY%/*}" | jq -r ".data.data.${PASSWORD_KEY##*/}")

> ID_KEY="$(yq -r '. | select(.kind == "Secret" and .metadata.name == "object-backup-s3-credentials").stringData.username' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"

Reviewer: ```suggestion
ID_KEY="$(yq e '. | select(.kind == "Secret" and .metadata.name == "object-backup-s3-credentials").stringData.username' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"
```

export AZURE_ACCOUNT_NAME=$(vault kv get -format json "clusters/kv/${ID_KEY%/*}" | jq -r ".data.data.${ID_KEY##*/}")

> SECRET_KEY="$(yq -r '. | select(.kind == "Secret" and .metadata.name == "object-backup-s3-credentials").stringData.password' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"

Reviewer: ```suggestion
SECRET_KEY="$(yq e '. | select(.kind == "Secret" and .metadata.name == "object-backup-s3-credentials").stringData.password' catalog/manifests/cluster-backup/05_schedule.yaml | cut -d: -f2)"
```
